src/MinresBlockSolver.cpp

```diff
@@ -20,311 +20,175 @@
+        SetPrintLevel(print_level_);
     }
 
-    minres_.Mult(*rhs_, sol);
+    nnz_ = M_.nnz() + DT_.nnz() + D_.nnz() + W_.nnz();
+}
 
-    chrono.Stop();
-    timing_ += chrono.RealTime();
 
-    if (myid_ == 0 && print_level_ > 0)
-    {
-        std::cout << "  Timing MINRES: Solver done in "
-                  << timing_ << "s. \n";
-        if (minres_.GetConverged())
-            std::cout << "  Minres converged in "
-                      << minres_.GetNumIterations()
-                      << " with a final residual norm "
-                      << minres_.GetFinalNorm() << "\n";
-        else
-            std::cout << "  Minres did not converge in "
-                      << minres_.GetNumIterations()
-                      << ". Final residual norm is "
-                      << minres_.GetFinalNorm() << "\n";
-    }
+MinresBlockSolver::MinresBlockSolver(const MinresBlockSolver& other) noexcept
+    : MGLSolver(other), op_(other.op_), prec_(other.prec_),
+      M_prec_(other.M_prec_), schur_prec_(other.schur_prec_),
+      pminres_(other.pminres_),
+      true_rhs_(other.true_rhs_), true_sol_(other.true_sol_)
+{
 
-    num_iterations_ += minres_.GetNumIterations();
 }
 
-/**
-   MinresBlockSolver acts on "true" dofs, this one does not.
-*/
-MinresBlockSolverFalse::MinresBlockSolverFalse(
-    MPI_Comm comm, const MixedMatrix& mgL)
-    :
-    MinresBlockSolver(comm, mgL),
-    mixed_matrix_(mgL),
-    true_rhs_(mgL.get_blockTrueOffsets()),
-    true_sol_(mgL.get_blockTrueOffsets())
+MinresBlockSolver::MinresBlockSolver(MinresBlockSolver&& other) noexcept
 {
+    swap(*this, other);
 }
 
-MinresBlockSolverFalse::~MinresBlockSolverFalse()
+MinresBlockSolver& MinresBlockSolver::operator=(MinresBlockSolver other) noexcept
 {
+    swap(*this, other);
+
+    return *this;
 }
 
-void MinresBlockSolverFalse::Mult(const mfem::BlockVector& rhs,
-                                  mfem::BlockVector& sol) const
+void swap(MinresBlockSolver& lhs, MinresBlockSolver& rhs) noexcept
 {
-    mfem::StopWatch chrono;
-    chrono.Clear();
-    chrono.Start();
+    swap(static_cast<MGLSolver&>(lhs),
+         static_cast<MGLSolver&>(rhs));
+
+    swap(lhs.op_, rhs.op_);
+    swap(lhs.prec_, rhs.prec_);
+    swap(lhs.M_prec_, rhs.M_prec_);
+    swap(lhs.schur_prec_, rhs.schur_prec_);
+    swap(lhs.pminres_, rhs.pminres_);
+    swap(lhs.true_rhs_, rhs.true_rhs_);
+    swap(lhs.true_sol_, rhs.true_sol_);
+}
 
-    const mfem::HypreParMatrix& edgedof_d_td = mixed_matrix_.get_edge_d_td();
+void MinresBlockSolver::Solve(const BlockVector& rhs, BlockVector& sol) const
+{
+    Timer timer(Timer::Start::True);
 
-    edgedof_d_td.MultTranspose(rhs.GetBlock(0), true_rhs_.GetBlock(0));
+    edge_true_edge_.MultAT(rhs.GetBlock(0), true_rhs_.GetBlock(0));
     true_rhs_.GetBlock(1) = rhs.GetBlock(1);
+    true_sol_ = 0.0;
 
-    if (!use_W_ && myid_ == 0)
+    if (!use_w_ && myid_ == 0)
     {
-        true_rhs_.GetBlock(1)(0) = 0.0;
+        true_rhs_[0] = 0.0;
```

Comment:
should here be `true_rhs_.GetBlock(1)[0]`?

In fact, I saw in the multilevel code this value does not affect the solution. I dont understand what is going on. But this may have something to do with "iterative mode"? In the original code, if minres' `iterative_mode` is false, this value will affect the solution, but when `iterative_mode` is true, similar behavior as in this new code. On the other hand, the 0-th entry of the final solution will equal to the 0-th entry of the initial guess.

Reply:
I noticed this too when implementing ML branch and it should be fixed now (in the `feedback-dev` branch anyway).